### ISSUE TITLE
Performance degradation

### ISSUE THREAD

**Issue by Jurigag** [zephir-lang/zephir#1507]
Im not telling that fixed issues are bad or something, but those method call time difference in microbench.php is pretty big:

```
Benchmark Zephir 0.9.6a-dev
empty for in range 0.048
$x = $this->x      0.262
$this->x = 0       0.261
self::a = 0        0.130
$this->x++         0.532
$this->x--         0.541
isset($this->x)    0.115
isset(self::A)     0.048
empty($this->x)    0.298
empty(self::a)     0.256
$this->f()         0.260
self::f()          0.266
self::f() -> true  0.262
$x = Foo::TEST     0.051
------------------------
Total              3.332
```

```
Benchmark Zephir 0.9.7
empty for in range 0.046
$x = $this->x      0.273
$this->x = 0       0.271
self::a = 0        0.230
$this->x++         0.537
$this->x--         0.547
isset($this->x)    0.113
isset(self::A)     0.046
empty($this->x)    0.307
empty(self::a)     0.249
$this->f()         1.534
self::f()          1.825
self::f() -> true  1.852
$x = Foo::TEST     0.052
------------------------
Total              7.883
```

Are there any plans to improve current one speed if it's anyhow possible?

Keep in mind that this is only about method calling itself, like how it's called(am i right?). Method in microbench is called 5000000, in real app on phalcon i guess there is less than 5000 calls.

**Comment by virgofx** [comment]
Eeek. Definitely :+1:

**Comment by Jurigag** [comment]
I'm guessing that we could partially improve it by changing some method in phalcon to internal, it shouldn't be affected by those changes, right? I would like even to suggest maybe something like change internal to allow it with combination of protected like:

```
protected internal function xyz()
```

And add zephir buildinternal - when build like this, then xyz method will be only internal.

@sjinks

**Comment by sjinks** [comment]
What PHP version did you use?

**Comment by Jurigag** [comment]
PHP 7, will check it on php 7.1

**Comment by sjinks** [comment]
0.9.6a-dev, what git revision was that?

Want to try to test locally.

**Comment by Jurigag** [comment]
It's 0.9.6 tag.

`$this->x += 2`

doesn't run on it, so keep this in mind.

**Comment by sergeyklay** [comment]
@Jurigag Could you please provide the git commit (SHA-1)?

**Comment by Jurigag** [comment]
https://github.com/phalcon/zephir/commits/0.9.6 https://github.com/phalcon/zephir/commit/72dbb2063e835ffa860dcf235dcbf6ec0f821fa5

**Comment by sjinks** [comment]
@Jurigag please take a look #1509

**Comment by sjinks** [comment]
It is still slower than 72dbb20:

72dbb20:
```
$this->f()         0.192
self::f()          0.189
self::f() -> true  0.193
```

#1509 
```
$this->f()         0.381
self::f()          0.468
self::f() -> true  0.467
```

but that's the price for correctness — the checks to make sure that the proper method is called in static/self contexts.

**Comment by Jurigag** [comment]
Well i can live with that difference, still big improvement. If it doesn't create any issues or problems i think we can merge it and maybe release new tag?

Im guessing that we will see in travis ci how faster tests will finish with this change.

**Comment by Jurigag** [comment]
@sergeyklay @sjinks don't forget about release, it could be good, at least phalcon in travis using latest stable now(release), right?

**Comment by sergeyklay** [comment]
I propose do release after closing this https://github.com/phalcon/zephir/pull/1479

**Comment by Jurigag** [comment]
Okay then.

**Comment by Jurigag** [comment]
@sjinks this should be reopened or not? The problem is still there and we need to find why tests failed after merging right? Or you did this already?

**Comment by sjinks** [comment]
I have fixed the 5.x thing, now looking at 7.x.

**Comment by sjinks** [comment]
Judging by https://github.com/phalcon/cphalcon/pull/12825, 7.x also works now.

**Comment by Jurigag** [comment]
Yea but fix was reverted after all or you readded it in some PR?

**Comment by sjinks** [comment]
#1519 if I am not mistaken

**Comment by Jurigag** [comment]
https://github.com/phalcon/zephir/commit/d90e076eeae0457923a075f454ba86a5edaefebe here is something - well let me just test and check it.

**Comment by Jurigag** [comment]
Well on latest zephir i got result like this:

```
empty for in range 0.049
$x = $this->x      0.273
$this->x = 0       0.267
self::a = 0        0.229
$this->x++         0.553
$this->x--         0.550
isset($this->x)    0.115
isset(self::A)     0.049
empty($this->x)    0.306
empty(self::a)     0.254
$this->f()         9.321
self::f()          9.848
self::f() -> true  9.699
$x = Foo::TEST     0.053
------------------------
Total             31.566
```

:C

So i guess we need to go back to this fix and find out why it was broken you reverted it here - https://github.com/phalcon/zephir/commit/5f18566712c677187bea07626a856457c2a0ded1

**Comment by sjinks** [comment]
Are you looking at Travis builds or do you run benchmark yourself?

```
Benchmark Zephir 0.9.7
empty for in range 0.100
$x = $this->x      0.828
$this->x = 0       0.846
self::a = 0        0.762
$this->x += 2      1.679
$this->x++         1.721
$this->x--         1.757
isset($this->x)    0.186
isset(self::A)     0.105
empty($this->x)    0.943
empty(self::a)     0.772
$this->f()         2.134
self::f()          2.282
self::f() -> true  2.292
$x = Foo::TEST     0.115
------------------------
Total             16.521
```

**Comment by Jurigag** [comment]
Im running benchmark myself.

**Comment by sjinks** [comment]
```
Benchmark Zephir 0.9.6a-dev
empty for in range 0.103
$x = $this->x      0.854
$this->x = 0       0.876
self::a = 0        0.410
$this->x++         1.743
$this->x--         1.742
isset($this->x)    0.189
isset(self::A)     0.093
empty($this->x)    0.932
empty(self::a)     0.819
$this->f()         1.369
self::f()          1.405
self::f() -> true  1.407
$x = Foo::TEST     0.116
------------------------
Total             12.057
```

**Comment by sjinks** [comment]
At that, I am running a debug build of PHP, release version is faster…

**Comment by Jurigag** [comment]
What php version? On 7.0.13 i have this 9 second result for some reason, `opcache.enable_cli = 1` i added

**Comment by sjinks** [comment]
PHP 7.0.17 (cli) (built: Apr  9 2017 20:37:39) ( ZTS DEBUG )

**Comment by sjinks** [comment]
Are we on the same master? a5f585482bf9c1f6a37d8162936765ebb3d9bfea

**Comment by Jurigag** [comment]
Oh sry, my mistake, forgot to disable xdebug:

```

Benchmark Zephir 0.9.7
empty for in range 0.051
$x = $this->x      0.280
$this->x = 0       0.277
self::a = 0        0.317
$this->x += 2      0.543
$this->x++         0.579
$this->x--         0.561
isset($this->x)    0.118
isset(self::A)     0.051
empty($this->x)    0.345
empty(self::a)     0.258
$this->f()         0.731
self::f()          0.827
self::f() -> true  0.924
$x = Foo::TEST     0.055
------------------------
Total              5.917
```

Looks like fixed. Slightly slower than 0.9.6 but so many fixes. Sorry for my mistake :C

**Comment by sjinks** [comment]
That's the price for #251 and #1429 :-(

**Comment by Jurigag** [comment]
Though in https://github.com/phalcon/zephir/pull/1509 https://github.com/phalcon/zephir/issues/1507#issuecomment-294210334 it was faster, not everything from there you finally added? Or some other fix from somewhere make it worse.

**Comment by sjinks** [comment]
#1522 slowed things down but just a bit.

**Comment by Jurigag** [comment]
Well at least it's better than it was before because almost 9x times performance degrndation in OP was kind of big, with around ~3 times i can live.

Also why now 

```
$this->f()         0.731
self::f()          0.827
self::f() -> true  0.924
```

This differ much? Previously(0.9.6) it was almost on same performance level.

**Comment by sjinks** [comment]
Are you building Zephir with `-O2` or `-O0`?

On our end there is almost no difference between `self::` and `$this->`, the difference comes from ZE.

**Comment by Jurigag** [comment]
Im just doing zephir build in zephir directory.

**Comment by sjinks** [comment]
What we add to `zend_call_function()` is `zephir_make_fcall_key()` and `populate_fcic()` — they add some overhead comparing to 061fec5d7bb6d0338d6295c1ed516dc0900ee51b

**Comment by sjinks** [comment]
Please try this patch, does it make any measurable difference for you?

```diff
diff --git a/kernels/ZendEngine3/fcall.c b/kernels/ZendEngine3/fcall.c
index 4bbc9a5..63940b0 100644
--- a/kernels/ZendEngine3/fcall.c
+++ b/kernels/ZendEngine3/fcall.c
@@ -47,9 +47,9 @@ int zephir_has_constructor_ce(const zend_class_entry *ce)
 /**
  * Creates a unique key to cache the current method/function call address for the current scope
  */
-static zend_string* zephir_make_fcall_key(zephir_call_type type, zend_class_entry *ce, zval *function)
+static zend_string* zephir_make_fcall_key(zephir_call_type type, zend_class_entry *ce, zval *function, zend_class_entry* called_scope)
 {
-       const zend_class_entry *calling_scope, *called_scope;
+       const zend_class_entry *calling_scope;
        zend_string *res = NULL;
        char *buf = NULL, *c;
        size_t l = 0, len = 0;
@@ -61,8 +61,6 @@ static zend_string* zephir_make_fcall_key(zephir_call_type type, zend_class_entr
        calling_scope = EG(scope);
 #endif

-       called_scope = zend_get_called_scope(EG(current_execute_data));
-
        switch (type) {
                case zephir_fcall_parent:
                        if (UNEXPECTED(!calling_scope || !calling_scope->parent)) {
@@ -209,7 +207,7 @@ static void resolve_callable(zval* retval, zephir_call_type type, zend_class_ent
        } ZEND_HASH_FILL_END();
 }

-static void populate_fcic(zend_fcall_info_cache* fcic, zephir_call_type type, zend_class_entry* ce, zval *this_ptr, zval *func)
+static void populate_fcic(zend_fcall_info_cache* fcic, zephir_call_type type, zend_class_entry* ce, zval *this_ptr, zval *func, zend_class_entry* called_scope)
 {
        zend_class_entry* calling_scope;
        fcic->initialized      = 0;
@@ -223,7 +221,7 @@ static void populate_fcic(zend_fcall_info_cache* fcic, zephir_call_type type, ze
                return;
        }

-       fcic->called_scope = zend_get_called_scope(EG(current_execute_data));
+       fcic->called_scope = called_scope;

 #if PHP_VERSION_ID >= 70100
        calling_scope = zend_get_executed_scope();
@@ -296,6 +294,7 @@ int zephir_call_user_function(zval *object_pp, zend_class_entry *obj_ce, zephir_
        zephir_fcall_cache_entry *temp_cache_entry = NULL;
        zval callable;
        int reload_cache = 1, i;
+       zend_class_entry* called_scope = zend_get_called_scope(EG(current_execute_data));

        assert(obj_ce || !object_pp);
        ZVAL_UNDEF(&callable);
@@ -323,7 +322,7 @@ int zephir_call_user_function(zval *object_pp, zend_class_entry *obj_ce, zephir_
                }

                if (reload_cache) {
-                       fcall_key = zephir_make_fcall_key(type, (object_pp && type != zephir_fcall_ce ? Z_OBJCE_P(object_pp) : obj_ce), function_name);
+                       fcall_key = zephir_make_fcall_key(type, (object_pp && type != zephir_fcall_ce ? Z_OBJCE_P(object_pp) : obj_ce), function_name, called_scope);
                        if (fcall_key) {
                                temp_cache_entry = zend_hash_find_ptr(zephir_globals_ptr->fcache, fcall_key);
                                if (temp_cache_entry) {
@@ -348,7 +347,7 @@ int zephir_call_user_function(zval *object_pp, zend_class_entry *obj_ce, zephir_
        fcic.initialized = 0;
        if (cache_entry && *cache_entry) {
        /* We have a cache record, initialize scope */
-               populate_fcic(&fcic, type, obj_ce, object_pp, function_name);
+               populate_fcic(&fcic, type, obj_ce, object_pp, function_name, called_scope);
                if (!fcic.function_handler) {
                        fcic.function_handler = *cache_entry;
                }
```

**Comment by Jurigag** [comment]
Yea, small difference:

```
empty for in range 0.054
$x = $this->x      0.271
$this->x = 0       0.276
self::a = 0        0.310
$this->x += 2      0.539
$this->x++         0.557
$this->x--         0.553
isset($this->x)    0.114
isset(self::A)     0.049
empty($this->x)    0.323
empty(self::a)     0.268
$this->f()         0.785
self::f()          0.795
self::f() -> true  0.824
$x = Foo::TEST     0.064
```

Chcked it more times and pretty much now on similar level those calls.

**Comment by Jurigag** [comment]
On latest zephir is someone intersted with latest PR https://github.com/phalcon/zephir/pull/1529 result is:

```
Benchmark Zephir 0.9.7
empty for in range 0.047
$x = $this->x      0.269
$this->x = 0       0.263
self::a = 0        0.229
$this->x += 2      0.545
$this->x++         0.540
$this->x--         0.548
isset($this->x)    0.111
isset(self::A)     0.048
empty($this->x)    0.315
empty(self::a)     0.238
$this->f()         0.361
self::f()          0.361
self::f() -> true  0.366
$x = Foo::TEST     0.052
------------------------
Total              4.293
```

So pretty much second difference but many fixes on latest zephir compared to 0.9.6-dev
